rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll

```diff
@@ -136,14 +136,40 @@ class LogicalAndBinaryOpExprTree extends PreOrderTree, LogicalAndExpr {
   }
 }
 
-class BlockExprBaseTree extends StandardPostOrderTree instanceof BlockExpr {
+class BlockExprTree extends StandardPostOrderTree, BlockExpr {
   override AstNode getChildNode(int i) {
     result = super.getStmtList().getStatement(i)
     or
     not exists(super.getStmtList().getStatement(i)) and
     (exists(super.getStmtList().getStatement(i - 1)) or i = 0) and
     result = super.getStmtList().getTailExpr()
   }
+
+  override predicate propagatesAbnormal(AstNode child) { none() }
+
+  /** Holds if this block captures the break completion `c`. */
+  private predicate capturesBreakCompletion(LoopJumpCompletion c) {
+    c.isBreak() and
```

Comment:
Do you also handle shadowing of labels? It's a silly thing to do, and the rust compiler will warn about it, but shadowed labels are allowed. For example:
```rust
        'a: loop {
            'a: loop {
                'a: {
                    break 'a;
                }
                break 'a;
            }
            break 'a;
        }
```
You might want to add a test case, if you don't already have one;

Reply:
There is already one: https://github.com/github/codeql/blob/main/rust/ql/test/library-tests/controlflow/test.rs#L26.

Reply:
Oh, wait, I we don't have one that shadows; I'll add.

Reply:
I have added one now, and the CFG is correct:

```mermaid
flowchart TD
1["enter test_loop_label_shadowing"]
10["b"]
11["ContinueExpr"]
12["ExprStmt"]
13["ContinueExpr"]
14["ExprStmt"]
2["1"]
3["ExprStmt"]
4["ExprStmt"]
5["IfExpr"]
6["b"]
7["ContinueExpr"]
8["ExprStmt"]
9["IfExpr"]

1 --> 3
2 --> 4
3 --> 2
4 --> 6
5 --> 14
6 -- false --> 10
6 -- true --> 8
7 -- continue --> 4
8 --> 7
9 --> 5
10 -- false --> 9
10 -- true --> 12
11 -- continue('loop) --> 4
12 --> 11
13 -- continue('loop) --> 4
14 --> 13
```